<a href="https://colab.research.google.com/github/MH-mirghaderi/DataScience.course/blob/main/AmazonSalesAnalysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<font face="XB Zar" size=5><div
style="margin:auto;" dir=rtl align=center>
<font face="IranNastaliq" size=5>
We Trust in God
</font>
<br>
<font face="IranNastaliq" size=5>
دانشگاه شهید بهشتی تهران - دانشکده علوم ریاضی (دپارتمان علوم کامپیوتر)
</font>
<br>
<font color=blue size=5>
مبانی علوم داده
<br>
<font color=lightgreen size=5>

بهار 1404 - دکتر خردپیشه
</font>
<br>
<hr/>
<font color=white size=5>
Data Science (Spring 2025)
</font>
<br>
Dr. Kheradpishe
</div>

#Amazon Sales Analysis

In [ ]:
import pandas as pd
import numpy as np
import scipy.stats as stats
import matplotlib.pyplot as plt
import seaborn as sns
df = pd.read_csv('/content/drive/Othercomputers/My Laptop/Desktop/Data Science/amazon.csv')


In [ ]:
df.head()

product_id                                       product_name  \
0  B07JW9H4J1  Wayona Nylon Braided USB to Lightning Fast Cha...   
1  B098NS6PVG  Ambrane Unbreakable 60W / 3A Fast Charging 1.5...   
2  B096MSW6CT  Sounce Fast Phone Charging Cable & Data Sync U...   
3  B08HDJ86NZ  boAt Deuce USB 300 2 in 1 Type-C & Micro USB S...   
4  B08CF3B7N1  Portronics Konnect L 1.2M Fast Charging 3A 8 P...   

                                            category discounted_price  \
0  Computers&Accessories|Accessories&Peripherals|...             ₹399   
1  Computers&Accessories|Accessories&Peripherals|...             ₹199   
2  Computers&Accessories|Accessories&Peripherals|...             ₹199   
3  Computers&Accessories|Accessories&Peripherals|...             ₹329   
4  Computers&Accessories|Accessories&Peripherals|...             ₹154   

  actual_price discount_percentage rating rating_count  \
0       ₹1,099                 64%    4.2       24,269   
1         ₹349                 43%    4.0       43,994   
2       ₹1,899                 90%    3.9        7,928   
3         ₹699                 53%    4.2       94,363   
4         ₹399                 61%    4.2       16,905   

                                       about_product  \
0  High Compatibility : Compatible With iPhone 12...   
1  Compatible with all Type C enabled devices, be...   
2  【 Fast Charger& Data Sync】-With built-in safet...   
3  The boAt Deuce USB 300 2 in 1 cable is compati...   
4  [CHARGE & SYNC FUNCTION]- This cable comes wit...   

                                             user_id  \
0  AG3D6O4STAQKAY2UVGEUV46KN35Q,AHMY5CWJMMK5BJRBB...   
1  AECPFYFQVRUWC3KGNLJIOREFP5LQ,AGYYVPDD7YG7FYNBX...   
2  AGU3BBQ2V2DDAMOAKGFAWDDQ6QHA,AESFLDV2PT363T2AQ...   
3  AEWAZDZZJLQUYVOVGBEUKSLXHQ5A,AG5HTSFRRE6NL3M5S...   
4  AE3Q6KSUK5P75D5HFYHCRAOLODSA,AFUGIFH5ZAFXRDSZH...   

                                           user_name  \
0  Manav,Adarsh gupta,Sundeep,S.Sayeed Ahmed,jasp...   
1  ArdKn,Nirbhay kumar,Sagar Viswanathan,Asp,Plac...   
2  Kunal,Himanshu,viswanath,sai niharka,saqib mal...   
3  Omkar dhale,JD,HEMALATHA,Ajwadh a.,amar singh ...   
4  rahuls6099,Swasat Borah,Ajay Wadke,Pranali,RVK...   

                                           review_id  \
0  R3HXWT0LRP0NMF,R2AJM3LFTLZHFO,R6AQJGUP6P86,R1K...   
1  RGIQEG07R9HS2,R1SMWZQ86XIN8U,R2J3Y1WL29GWDE,RY...   
2  R3J3EQQ9TZI5ZJ,R3E7WBGK7ID0KV,RWU79XKQ6I1QF,R2...   
3  R3EEUZKKK9J36I,R3HJVYCLYOY554,REDECAZ7AMPQC,R1...   
4  R1BP4L2HH9TFUP,R16PVJEXKV6QZS,R2UPDB81N66T4P,R...   

                                        review_title  \
0  Satisfied,Charging is really fast,Value for mo...   
1  A Good Braided Cable for Your Type C Device,Go...   
2  Good speed for earlier versions,Good Product,W...   
3  Good product,Good one,Nice,Really nice product...   
4  As good as original,Decent,Good one for second...   

                                      review_content  \
0  Looks durable Charging is fine tooNo complains...   
1  I ordered this cable to connect my phone to An...   
2  Not quite durable and sturdy,https://m.media-a...   
3  Good product,long wire,Charges good,Nice,I bou...   
4  Bought this instead of original apple, does th...   

                                            img_link  \
0  https://m.media-amazon.com/images/W/WEBP_40237...   
1  https://m.media-amazon.com/images/W/WEBP_40237...   
2  https://m.media-amazon.com/images/W/WEBP_40237...   
3  https://m.media-amazon.com/images/I/41V5FtEWPk...   
4  https://m.media-amazon.com/images/W/WEBP_40237...   

                                        product_link  
0  https://www.amazon.in/Wayona-Braided-WN3LG1-Sy...  
1  https://www.amazon.in/Ambrane-Unbreakable-Char...  
2  https://www.amazon.in/Sounce-iPhone-Charging-C...  
3  https://www.amazon.in/Deuce-300-Resistant-Tang...  
4  https://www.amazon.in/Portronics-Konnect-POR-1...

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1465 entries, 0 to 1464
Data columns (total 16 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   product_id           1465 non-null   object
 1   product_name         1465 non-null   object
 2   category             1465 non-null   object
 3   discounted_price     1465 non-null   object
 4   actual_price         1465 non-null   object
 5   discount_percentage  1465 non-null   object
 6   rating               1465 non-null   object
 7   rating_count         1463 non-null   object
 8   about_product        1465 non-null   object
 9   user_id              1465 non-null   object
 10  user_name            1465 non-null   object
 11  review_id            1465 non-null   object
 12  review_title         1465 non-null   object
 13  review_content       1465 non-null   object
 14  img_link             1465 non-null   object
 15  product_link         1465 non-null   object
dtypes: obj

In [ ]:
df.describe()

product_id                                       product_name  \
count         1465                                               1465   
unique        1351                                               1337   
top     B083342NKJ  Fire-Boltt Ninja Call Pro Plus 1.83" Smart Wat...   
freq             3                                                  5   

                                                 category discounted_price  \
count                                                1465             1465   
unique                                                211              550   
top     Computers&Accessories|Accessories&Peripherals|...             ₹199   
freq                                                  233               53   

       actual_price discount_percentage rating rating_count  \
count          1465                1465   1465         1463   
unique          449                  92     28         1143   
top            ₹999                 50%    4.1        9,378   
freq            120                  56    244            9   

                                            about_product  \
count                                                1465   
unique                                               1293   
top     [CHARGE & SYNC FUNCTION]- This cable comes wit...   
freq                                                    6   

                                                  user_id  \
count                                                1465   
unique                                               1194   
top     AHIKJUDTVJ4T6DV6IUGFYZ5LXMPA,AE55KTFVNXYFD5FPY...   
freq                                                   10   

                                                user_name  \
count                                                1465   
unique                                               1194   
top     $@|\|TO$|-|,Sethu madhav,Akash Thakur,Burger P...   
freq                                                   10   

                                                review_id  \
count                                                1465   
unique                                               1194   
top     R3F4T5TRYPTMIG,R3DQIEC603E7AY,R1O4Z15FD40PV5,R...   
freq                                                   10   

                                             review_title  \
count                                                1465   
unique                                               1194   
top     Worked on iPhone 7 and didn’t work on XR,Good ...   
freq                                                   10   

                                           review_content  \
count                                                1465   
unique                                               1212   
top     I am not big on camera usage, personally. I wa...   
freq                                                    8   

                                                 img_link  \
count                                                1465   
unique                                               1412   
top     https://m.media-amazon.com/images/I/413sCRKobN...   
freq                                                    3   

                                             product_link  
count                                                1465  
unique                                               1465  
top     https://www.amazon.in/Borosil-Jumbo-1000-Watt-...  
freq                                                    1

In [ ]:
df.isna().sum()

product_id             0
product_name           0
category               0
discounted_price       0
actual_price           0
discount_percentage    0
rating                 0
rating_count           2
about_product          0
user_id                0
user_name              0
review_id              0
review_title           0
review_content         0
img_link               0
product_link           0
dtype: int64

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1465 entries, 0 to 1464
Data columns (total 16 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   product_id           1465 non-null   object
 1   product_name         1465 non-null   object
 2   category             1465 non-null   object
 3   discounted_price     1465 non-null   object
 4   actual_price         1465 non-null   object
 5   discount_percentage  1465 non-null   object
 6   rating               1465 non-null   object
 7   rating_count         1463 non-null   object
 8   about_product        1465 non-null   object
 9   user_id              1465 non-null   object
 10  user_name            1465 non-null   object
 11  review_id            1465 non-null   object
 12  review_title         1465 non-null   object
 13  review_content       1465 non-null   object
 14  img_link             1465 non-null   object
 15  product_link         1465 non-null   object
dtypes: obj

In [ ]:
import pandas as pd

cols = ['discounted_price', 'actual_price', 'discount_percentage','rating_count','rating']

def clean_and_convert(s):
    return (
        s.astype(str)
         # ۱. حذف فاصله‌های غیرمعمول (از جمله ZWNJ)
         .str.replace(r'[\u200c\s]', '', regex=True)
         # ۲. حذف جداکننده‌های هزارگان انگلیسی و فارسی
         .str.replace(r'[٬,]', '', regex=True)
         # ۳. حذف همه‌چیز به‌جز ارقام و نقطه
         .str.replace(r'[^0-9\.]', '', regex=True)
         # ۴. تبدیل نهایی به عدد (خطاها => NaN)
         .pipe(pd.to_numeric, errors='coerce')
    )

for col in cols:
    df[col] = clean_and_convert(df[col])


In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1465 entries, 0 to 1464
Data columns (total 16 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   product_id           1465 non-null   object 
 1   product_name         1465 non-null   object 
 2   category             1465 non-null   object 
 3   discounted_price     1465 non-null   float64
 4   actual_price         1465 non-null   float64
 5   discount_percentage  1465 non-null   int64  
 6   rating               1464 non-null   float64
 7   rating_count         1463 non-null   float64
 8   about_product        1465 non-null   object 
 9   user_id              1465 non-null   object 
 10  user_name            1465 non-null   object 
 11  review_id            1465 non-null   object 
 12  review_title         1465 non-null   object 
 13  review_content       1465 non-null   object 
 14  img_link             1465 non-null   object 
 15  product_link         1465 non-null   o

In [ ]:
df=df.dropna(subset=['rating_count','rating'])

In [ ]:
df.isna().sum()

product_id             0
product_name           0
category               0
discounted_price       0
actual_price           0
discount_percentage    0
rating                 0
rating_count           0
about_product          0
user_id                0
user_name              0
review_id              0
review_title           0
review_content         0
img_link               0
product_link           0
dtype: int64

#1 Statistical Questions


##1. Does the discounted price significantly impact product ratings?

In [ ]:
from scipy.stats import spearmanr
rho , pval = spearmanr(df['discounted_price'] , df['rating'])
print(f'spearmans ρ = {rho:.4f} , p-value = {pval:.4f}' )

spearmans ρ = 0.0801 , p-value = 0.0022


### Result

>

1.   ρ (rho):
                this is spearman's rank-correlation coeficient.
  its value ranges from -1 to +1:


*   +1 means the ranks move perfectly together in the same (increasing) direction (perfect positive monotonic assosiation)
*   -1 means the ranks move perfectly in opposite direction(perfect negative monotonic assosiation)
*   0 means there's no clear rank-order relationship.

2.   p-value


*   this is the probability that, under the null hypothesis of 'no associaton' between the variables
*   if the p-value is below our chosen significance level(commonly 0.05), we reject the null and conclude that the observed correlation is statistically signicant.

Result : ρ = 0.0801 indicates a very,very weak association.
According to points that denoted above , actually no real association.






##2. Are product categories and high/low ratings independent?

In [ ]:
from scipy.stats import chisquare,chi2_contingency

In [ ]:
df['rating-level'] = np.where(df['rating'] >=4, 'High', 'Low')

<ipython-input-32-4acddd63a486>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['rating-level'] = np.where(df['rating'] >=4, 'High', 'Low')


In [ ]:
df['Main_Category'] = df['category'].str.split('|').str[0]

<ipython-input-33-c868dd9a4f38>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Main_Category'] = df['category'].str.split('|').str[0]


In [ ]:
contingency = pd.crosstab(df['Main_Category'], df['rating-level'],margins = True)
display(contingency)

rating-level           High  Low   All
Main_Category                         
Car&Motorbike             0    1     1
Computers&Accessories   377   74   451
Electronics             393  133   526
Health&PersonalCare       1    0     1
Home&Kitchen            303  144   447
HomeImprovement           2    0     2
MusicalInstruments        1    1     2
OfficeProducts           31    0    31
Toys&Games                1    0     1
All                    1109  353  1462

In [ ]:
chi2, p, dof, expected = chi2_contingency(contingency)
print(f"Chi2 statistic: {chi2:.2f}")
print(f"p-value: {p:.3e}")
print(f"Degrees of freedom: {dof}")


Chi2 statistic: 46.02
p-value: 2.947e-04
Degrees of freedom: 18




###Result
Since p < 0.05, we reject the null hypothesis of independence.
There is a statistically significant relationship between product category and whether a product receives a high (≥ 4) or low (< 4) rating.



##3. Is there a significant difference in ratings between high-discount and lowdiscount products?


In [ ]:
df['discount_group'] = df['discount_percentage'].apply(lambda x: 'High' if x > 50 else 'Low')
high = df[df['discount_group'] == 'High']['rating']
low = df[df['discount_group'] == 'Low']['rating']
tstat, pval = stats.ttest_ind(high, low, equal_var=False)
print("T-statistic:", tstat)
print("P-value:", pval)

T-statistic: -4.214820359775128
P-value: 2.6620121753262284e-05


<ipython-input-36-7efc67e94ccf>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['discount_group'] = df['discount_percentage'].apply(lambda x: 'High' if x > 50 else 'Low')


###Result
The average score between the two groups differs significantly. Products with high and low discounts are not comparable in terms of scores, and this difference is significant.

##4. Do more expensive products receive higher ratings on average?

In [ ]:
df['price_group'] = pd.qcut(df['actual_price'], 3, labels=['Low', 'Medium', 'High'])
groups = [group['rating'].dropna() for name, group in df.groupby('price_group')]
anova_stat, pval = stats.f_oneway(*groups)
print("ANOVA F-statistic:", anova_stat)
print("P-value:", pval)

ANOVA F-statistic: 0.9176997703336051
P-value: 0.3996672201025875


<ipython-input-37-3057a0c8e0db>:2: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  groups = [group['rating'].dropna() for name, group in df.groupby('price_group')]


###Result
There is no significant difference in scores among the price groups (low, medium, high). Therefore, it cannot be said that more expensive products necessarily receive higher scores.

##5. Does the distribution of rating counts follow a normal distribution?

In [ ]:
counts = df['rating_count'].dropna()
ks_stat, pval = stats.kstest((counts - counts.mean()) / counts.std(), 'norm')
print("K-S Statistic:", ks_stat)
print("P-value:", pval)

K-S Statistic: 0.3343128982515293
P-value: 3.4888965453650775e-146


###Result
The number of scores does not follow a normal distribution, and this difference is statistically highly significant. Therefore, nonparametric tests should be used for this variable.